### ISSUE TITLE
Slow training speed

### ISSUE THREAD

**Issue by vankhoa21991** [MIC-DKFZ/nnDetection#11]
Hello, I'm trying to train the task 16 with Luna dataset, but the training seems very slow, and the GPU is not used for most of the time because it has to wait for the volume to be prepared. How can I increase the volume preparation speed to take advantage of the GPU usage?

**Comment by mibaumgartner** [comment]
Hi @vankhoa21991 ,

there is no straightforward way to speed up the training while reproducing our results. Furthermore, it is necessary to find the bottleneck in the pipeline to make any changes.

The unpacked data should be located on an SSD (~500mb/s read definitely works, a little bit less should also work) to avoid IO bottlenecks. (this is visible by looking at the CPU util. if there is no load it is an IO bottleneck).

Adjust `det_num_threads` (similar to num workers in the normal pytorch dataloaders) for the available CPU resources (set this as high as possible but not more than available CPU threads) otherwise, there might be a CPU bottleneck.
- CPUs with less cores but high clock speed: Needs a lower `det_num_threads` value. On a Intel i& 9700 (non k) I use `det_num_threads=6` while retaining 90+ % GPU usage.
- CPUs with many cores but lower clock speed: Needs a high `det_num_threads` value. In our cluster environments with `det_num_threads=12` nnDet reached ~80+% GPU usage. 

If there are no additional CPU resources and it is really a CPU bottleneck it is possible to use different augmentation settings but these will reduce the final results.

Best,
Michael

**Comment by vankhoa21991** [comment]
Thanks, normally when you train on this dataset with default configuration, how long does it takes for one epoch? As I'm running it takes around 12 hours for one epoch.

**Comment by mibaumgartner** [comment]
The whole training (60 Epochs) takes around 2 days, so one epoch should take around 48 minutes.

**Comment by vankhoa21991** [comment]
Thanks, the reason maybe IO bound, but when I trained the model with nnunet on current hardware, it seems faster for each iteration, maybe the patch size is different, but gpu usage seems better.

**Comment by mibaumgartner** [comment]
There are some differences compared to nnUNet:

- Most 3D nnUNet configurations are trained with batch size 2. nnDetection uses batch size 4 and has a smaller network (batch size 2 didn't work well on some data sets). This results in more IO and CPU usage compared to nnUNet.
- nnDetection is slower per iteration (due to detection specific computations and the higher batch size) but uses significantly fewer iterations (around 60% of nnUNets iterations). So comparing the time per iterations will give a wrong sense of the whole training time.

**Comment by Derekabc** [comment]
Hi, @vankhoa21991, how long does it take to train for one epoch now? It takes me about 2 hours to train one epoch on LUNA16 dataset with 2080 Ti

**Comment by vankhoa21991** [comment]
It takes me 30 mins for one epoch, you need to store data on SSD
